docs/architecture/adr-054-protobuf-semver-compat-codegen.md

```diff
@@ -0,0 +1,248 @@
+a handler approach. It is a complex refactoring especially of `ValidateBasic`
+and alignment on an ideal replacement pattern hasn't been reached yet
+(https://github.com/cosmos/cosmos-sdk/pull/11340).
+
+#### Potential limitations to generated code
+
+In doing benchmarking of ORM generated code several potential optimizations have emerged and obviously the most
+performant optimization would be to generate as much code as possible. But because ORM table descriptors may evolve from
+one version to the next with new indexes and fields being added, this makes generated code state machine breaking.
+Recall that the API module paradigm requires no state machine breaking logic in generated code.
+
+## Decision
+
+To address these issues, we will adopt a hybrid approach which requires changes
+to the https://github.com/cosmos/cosmos-proto code generator.
+
+For client facing code, the API module pattern is still useful given that
+care is taken to only publish stable tags when packages are actually stable
+and in production.
+
+For state machine code, the recommended solution is to generate protobuf
+code used by the module in an `internal` package and modify the code generator
+to *skip global registration* for generated code in internal packages.
+
+Because there are plenty of cases where we want one module to essentially
+be a client of another module, either in `Keeper` method or as proposed in
+[ADR 033](./adr-033-protobuf-inter-module-comm.md), modules will need
+to use different generated code - for instance from an API module - than
+the code used in the state machine itself.
+
+To address this, we will also modify the code generator to generate interfaces
+which allow different concrete structs to be passed into keeper methods and
+ADR 033 than those used internally by the module.
+
+For instance, for a struct `MsgSend`, getters like `GetAmount`, `GetFromAddress`, are already defined so codegen could
+create an interface:
+
+```go
+type MsgSendI interface {
+  protoreflect.ProtoMessage
+  GetFromAddress() string
+  GetToAddress() string
+  GetAmount() []v1beta1.CoinI
+  
+  IsCosmosBankV1MsgSend() // this method distinguishes this interface from other proto types which may otherwise implement the same fields.
+}
+```
+
+The gRPC generated server and client could look like this, with the client having
+a way to detect the server revision (as described
+in https://github.com/cosmos/cosmos-sdk/blob/fdd3d07a28f662c2fabd6007a4a1f64be3a373b3/proto/cosmos/app/v1alpha1/module.proto#L55)
+to see which features the server supports if the client is newer:
+
+```go
+type MsgServer interface {
+  Send(context.Context, MsgSendI) (MsgSendResponseI, error)
+}
+
+type MsgClient interface {
+  Send(ctx context.Context, in MsgSendI, opts …grpc.CallOption) (MsgSendResponseI, error)
+  // note that for MsgSendResponseI we would need to have a generated response wrapper type to deal with the case where
```

Comment:
Is there a distinction in this ADR that describes client vs server version? Or is that something that already exists in the Proto usage within the SDK.

Reply:
What is meant by client and server version is simply the version of the proto files that the client and server have. It could be the scenario that the client has newer proto files than the server and vice versa. We need to gracefully handle both.